core/dbt/parser/parsing-vs-compilation-vs-runtime.md

```diff
@@ -0,0 +1,151 @@
+
+</details>
+
+## What is "execution"?
+
+**In a sentence:** Now that dbt knows about all the stuff in your project, it can perform operations on top of it.
+
+Things it can do:
+
+- tell you about all the models that match certain criteria (`list`)
+- compile + run a set of models, in DAG order
+- interactively compile / preview some Jinja-SQL, that includes calls to macros or ref’s models defined in your project
+
+Depending on what’s involved, these operations may or may not require a live database connection. While executing, dbt produces metadata, which it returns as **log events** and **artifacts**.
+
+Put another way, dbt’s execution has required inputs, expected outputs, and the possibility for side effects:
+
+- **Inputs** (provided by user): project files, credentials, configuration → Manifest + runtime configuration
+- **Outputs** (returned to user): logs & artifacts
+- **Side effects** (not seen directly by user): changes in database state, depending on the operation being performed
+
+### Compiling a model
+
+We use the word "compiling" in a way that’s confusing for most software engineers (and many other people). Most of what’s described above, parsing + validating + constructing a Manifest (internal representation), falls more squarely in the traditional role of a language compiler. By contrast, when we talk about "compiling SQL," we’re really talking about something that happens at **runtime**.
+
+Devils in the details; toggle away.
+
+<details>
+<summary>The mechanism of "compilation" varies by model language.</summary>
+
+- **Jinja-SQL** wants to compile down to "vanilla" SQL, appropriate for this database, where any calls to `ref('something')` have been replaced with `database.schema.something`.
+- dbt doesn’t directly modify or rewrite user-provided **Python** code at all. Instead, "compilation" looks like code generation: appending more methods that allow calls to `dbt.ref()`, `dbt.source()`, and `dbt.config.get()` to return the correct results at runtime.
+
+</details>
+
+<details>
+<summary>If your model’s code uses a dynamic query to template code, this requires a database connection.</summary>
+
+- At this point, `[execute](https://docs.getdbt.com/reference/dbt-jinja-functions/execute)` is set to `True`.
+- e.g. `dbt_utils.get_column_values`, `dbt_utils.star`
+- Jinja-SQL supports this sort of dynamic templating. Python does not; there are other imperative ways to do this, using DataFrame methods / the Python interpreter at runtime.
+
+</details>
+
+<details>
+<summary>Compilation is also when ephemeral model CTEs are interpolated into the models that `ref` them.</summary>
+
+- The code for this is *gnarly*. That’s all I’m going to say about it for now.
+
+</details>
+
+<details>
+<summary>When compiling happens for a given node varies by command.</summary>
+
+- In `dbt compile`, every model is compiled concurrently, up to the number of threads, rather than in DAG order.
```

Comment:
Since I wrote this, I figured out that it isn't actually how it works today. It _should_ work this way:
- https://github.com/dbt-labs/dbt-core/issues/7914

Let's cut the bullet for now